pkg/tenant/projects.go

```diff
@@ -514,3 +517,96 @@ func (s *Server) createProject(ctx context.Context, project *db.Project) (err er
 
 	return nil
 }
+
+// UpdateProjectStats updates the stat fields on a project by performing readonly
+// queries to Quarterdeck and Ensign. Because this requires a few RPCs, it should be
+// called in a background task where possible to avoid blocking user requests. The
+// context passed to this method must contain authentication credentials in order to
+// query Quarterdeck which must include the topics:read and projects:read permissions.
+// TODO: This data can be updated asynchronously once the Ensign "meta" topics are up
+// and running.
+func (s *Server) UpdateProjectStats(ctx context.Context, projectID ulid.ULID) (err error) {
+	// Go routine to fetch the number of project API keys from Quarterdeck.
+	countAPIKeys := func(ctx context.Context) (_ uint64, err error) {
+		var reply *qd.Project
+		if reply, err = s.quarterdeck.ProjectDetail(ctx, projectID.String()); err != nil {
+			return 0, err
+		}
+
+		return uint64(reply.APIKeysCount), nil
+	}
+
+	// Go routine to fetch the number of topics from Ensign.
+	countTopics := func(ctx context.Context) (_ uint64, err error) {
+		// Request access to the Ensign project from Quarterdeck.
+		req := &qd.Project{
+			ProjectID: projectID,
+		}
+
+		var reply *qd.LoginReply
+		if reply, err = s.quarterdeck.ProjectAccess(ctx, req); err != nil {
+			return 0, err
+		}
+
+		// Create special context with the one-time claims to make the request.
+		ensignContext := qd.ContextWithToken(ctx, reply.AccessToken)
+		info := &pb.InfoRequest{
+			Topics: [][]byte{[]byte(projectID[:])},
+		}
+
+		var project *pb.ProjectInfo
+		if project, err = s.ensign.Info(ensignContext, info); err != nil {
+			return 0, err
+		}
+
+		// Count the number of non-readonly topics.
+		if project.ReadonlyTopics > project.Topics {
+			return 0, nil
+		}
+
+		return project.Topics - project.ReadonlyTopics, nil
+	}
+
+	// Run both RPCs in parallel and capture errors.
+	var wg sync.WaitGroup
+	numAPIKeys, numTopics := new(uint64), new(uint64)
+	errs := make([]error, 2)
+	rpc := func(fn func(context.Context) (uint64, error), i int, res *uint64) {
+		defer wg.Done()
+		*res, errs[i] = fn(ctx)
+	}
+	wg.Add(2)
+	go rpc(countAPIKeys, 0, numAPIKeys)
+	go rpc(countTopics, 1, numTopics)
+	wg.Wait()
+
+	// Return if both RPCs errored.
```

Comment:
Should we return if only one of them errors to keep the Tenant view more consistent?

Reply:
Yes, you're probably right; e.g. if the user doesn't have read:topics we still want to know how many api keys.